### PR TITLE
Add timeout to file locking around saving and restoring caches.

### PR DESCRIPTION
**Pull request by shanejbrown**
Add timeout to file locking around saving and restoring caches. Add acquire and release lock functions in utils.py

## Motivation and Context

Some Jenkins jobs appear to be deadlocked around the existing file lock with caching. This is meant to lessen the pain of when deadlocks occur by allow other jobs to run. This doesn't address the root cause of the deadlocks however.

## How Has This Been Tested?

Unit tests were added around the new file lock timeout functions in utils

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: buildrunner/docker/runner.py [hunk 1]
```diff
@@ -323,32 +328,33 @@ def restore_caches(self, logger: ContainerLogger, caches: OrderedDict) -> None:
...
-                    finally:
-                        # Always unlock the file after extracting the archive
-                        fcntl.lockf(data, fcntl.LOCK_UN)
+                # Allow multiple people to read from the file at the same time
+                # If another instance adds an exclusive lock in the save method below,
+                # this will block until the lock is released
+                file_obj = acquire_read_binary_flock(
+                    lock_file=actual_cache_archive_file,
+                    logger=logger,
+                    timeout_seconds=600
```
Comment (bluesliverx) [inline buildrunner/docker/runner.py:344]: After thinking on this more, I think 10 minutes is pretty long. Maybe we should do 5?

File: tests/test_utils_flock.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            time.sleep(1)
+            fd = acquire_write_binary_flock(lock_file, mock_logger, timeout_seconds=2.0)
+            assert fd is not None
+            p.join()
+        finally:
+            if fd:
+                release_flock(fd, mock_logger)
+
+def test_flock_aquire_exlusive_timeout(mock_logger):
+    with tempfile.TemporaryDirectory() as tmp_dir_name:
```
Comment (bluesliverx) [inline tests/test_utils_flock.py:96]: Personally, I'd make this tmp_dir_name a fixture to cleanup this test code, but it's fine as is.

File: buildrunner/utils.py [hunk 3]
```diff
@@ -20,6 +23,13 @@
 
 from buildrunner.errors import BuildRunnerConfigurationError
 
+LOCK_TIMEOUT_SECONDS = 600.0
+
+class FailureToAcquireLockException(Exception):
```
Comment (bluesliverx) [inline buildrunner/utils.py:28]: This seems dumb, but there should be two spaces before a class or a method. This applies to everything below as well. pycodestyle usually catches this...

File: buildrunner/utils.py [hunk 4]
```diff
@@ -309,3 +319,118 @@ def _cycle_colors(colors):
...
+    :param mode: open mode
+    :param timeout_seconds: number of seconds for timeout
+    :param exclusive: config exclusive lock (True) or shared lock (False), defaults to True
+    :return: opened file object if successful else None
+    """
+    # Inspired by https://gist.github.com/jirihnidek/430d45c54311661b47fb45a3a7846537
+    flags = os.O_RDWR | os.O_CREAT
+
+    if 'a' in mode:
+        flags = flags | os.O_TRUNC
```
Comment (bluesliverx) [inline buildrunner/utils.py:343]: Truncate the file with an append? That doesn't make much sense to me.

File: buildrunner/utils.py [hunk 5]
```diff
@@ -309,3 +319,118 @@ def _cycle_colors(colors):
...
+            else:
+                # Allow multiple people to read from the file at the same time
+                # If another instance adds an exclusive lock in the save method below,
+                # this will block until the lock is released
+                fcntl.flock(file_obj, fcntl.LOCK_SH | fcntl.LOCK_NB)
+        except (IOError, OSError):
+            # Limit the amount of logs written while waiting for the file lock
+            num_seconds_between_log_writes = 10
+            if 0 <= (duration_seconds % num_seconds_between_log_writes) <= retry_sleep_seconds:
+                logger.write(f"PID:{pid} waiting for file lock on {lock_file} after {duration_seconds} seconds")
```
Comment (bluesliverx) [inline buildrunner/utils.py:371]: Nice.

File: buildrunner/docker/runner.py (hunk 1 above)
Comment (shanejbrown) [reply buildrunner/docker/runner.py:344]: Agreed. I'll make the change.

File: buildrunner/utils.py (hunk 3 above)
Comment (shanejbrown) [reply buildrunner/utils.py:28]: I will make that change. There is no `pycodestyle` in `buildrunner`.  I would add it in this PR but there are lots of places that would need to be fixed. So if we want to add it it should be a separate PR IMO. I'll make a ticket and we can discuss it.

File: buildrunner/utils.py (hunk 4 above)
Comment (shanejbrown) [reply buildrunner/utils.py:343]: Yeah I agree. I am not sure what I was thinking or why I have that there. I am going to relook at that function and the rest of this PR before this is merged.

File: tests/test_utils_flock.py (hunk 2 above)
Comment (shanejbrown) [reply tests/test_utils_flock.py:96]: Yeah I should use fixtures more. But if you are ok with it then I will probably just leave it as is.

File: buildrunner/docker/runner.py [hunk 6]
```diff
@@ -300,6 +315,11 @@ def _get_cache_file_from_prefix(
 
         return local_cache_archive_match
 
+    @timeout_decorator.timeout(CACHE_TIMEOUT_SECONDS, BuildRunnerCacheTimeout)
+    def _put_cache_in_container(self, restored_cache_src, docker_path, file_obj):
```
Comment (bluesliverx) [inline buildrunner/docker/runner.py:319]: Can you add type hints for all of the arguments here? Also a return type hint.

File: buildrunner/docker/runner.py [hunk 7]
```diff
@@ -323,32 +343,40 @@ def restore_caches(self, logger: ContainerLogger, caches: OrderedDict) -> None:
...
-                                f"WARNING: An error occurred when trying to use cache "
-                                f"{actual_cache_archive_file} at the path {docker_path}\n"
-                            )
-                    finally:
-                        # Always unlock the file after extracting the archive
-                        fcntl.lockf(data, fcntl.LOCK_UN)
+                # Allow multiple people to read from the file at the same time
+                # If another instance adds an exclusive lock in the save method below,
+                # this will block until the lock is released
+                file_obj = acquire_flock_open_read_binary_(
```
Comment (bluesliverx) [inline buildrunner/docker/runner.py:356]: Why the trailing underscore in the method name?

File: buildrunner/utils.py (hunk 3 above)
Comment (bluesliverx) [reply buildrunner/utils.py:29]: Agreed, thanks

File: buildrunner/docker/runner.py (hunk 7 above)
Comment (shanejbrown) [reply buildrunner/docker/runner.py]: Fixed

File: buildrunner/docker/runner.py (hunk 6 above)
Comment (shanejbrown) [reply buildrunner/docker/runner.py]: Done

**Review by bluesliverx — APPROVED** [review]

**Pull request merged**